extensions/amp-vk/0.1/test/test-amp-vk.js

```diff
@@ -0,0 +1,217 @@
+  });
+
+  function createAmpVkElement(dataParams, layout) {
+    const ele = doc.createElement('amp-vk');
+
+    for (const param in dataParams) {
+      ele.setAttribute(`data-${param}`, dataParams[param]);
+    }
+
+    ele.setAttribute('width', 500);
+    ele.setAttribute('height', 300);
+
+    if (layout) {
+      ele.setAttribute('layout', layout);
+    }
+
+    doc.body.appendChild(ele);
+
+    return ele.build().then(() => {
+      return ele.layoutCallback();
+    }).then(() => ele);
+  }
+
+  it('requires data-embedtype', () => {
+    const params = Object.assign({}, POST_PARAMS);
+    delete params['embedtype'];
+    return createAmpVkElement(params).should.eventually.be.rejectedWith(
+        /The data-embedtype attribute is required for/);
+  });
+
+  it('removes iframe after unlayoutCallback', () => {
+    return createAmpVkElement(POST_PARAMS).then(vkPost => {
+      const iframe = vkPost.querySelector('iframe');
+      expect(iframe).to.not.be.null;
+      const obj = vkPost.implementation_;
+      obj.unlayoutCallback();
+      expect(vkPost.querySelector('iframe')).to.be.null;
+      expect(obj.iframe_).to.be.null;
+      expect(obj.unlayoutOnPause()).to.be.true;
+    });
+  });
+
+  // Post tests
+
+  it('post::requires data-hash', () => {
+    const params = Object.assign({}, POST_PARAMS);
+    delete params['hash'];
+    return createAmpVkElement(params).should.eventually.be.rejectedWith(
+        /The data-hash attribute is required for/);
+  });
+
+  it('post::requires data-owner_id', () => {
+    const params = Object.assign({}, POST_PARAMS);
+    delete params['owner_id'];
+    return createAmpVkElement(params).should.eventually.be.rejectedWith(
+        /The data-owner_id attribute is required for/);
+  });
+
+  it('post::requires data-post_id', () => {
+    const params = Object.assign({}, POST_PARAMS);
+    delete params['post_id'];
+    return createAmpVkElement(params).should.eventually.be.rejectedWith(
+        /The data-post_id attribute is required for/);
+  });
+
+  it('post::renders iframe in amp-vk', () => {
+    return createAmpVkElement(POST_PARAMS).then(vkPost => {
+      const iframe = vkPost.querySelector('iframe');
+      expect(iframe).to.not.be.null;
+    });
+  });
+
+  it('post::renders responsively', () => {
+    return createAmpVkElement(POST_PARAMS, Layout.RESPONSIVE).then(vkPost => {
+      const iframe = vkPost.querySelector('iframe');
+      expect(iframe).to.not.be.null;
+      expect(iframe.className).to.match(/i-amphtml-fill-content/);
+    });
+  });
+
+  it('post::sets correct src url to the vk iFrame', () => {
+    return createAmpVkElement(POST_PARAMS, Layout.RESPONSIVE).then(vkPost => {
+      const iframe = vkPost.querySelector('iframe');
+      const referrer = encodeURIComponent(vkPost.ownerDocument.referrer);
+      const url = encodeURIComponent(
+          vkPost.ownerDocument.location.href.replace(/#.*$/, '')
+      );
+      const startWidth = iframe.offsetWidth;
```

Comment:
`this.getLayoutWidth()`

Reply:
fixed, however had to use
`const resource = Resource.forElement(element);`
`resource.measure();`
before layout callback and trigger `onLayoutMeasure();` before requesting layout width